File: DotNet/CesiumLanguageWriterTests/TestYearMonthDay.cs
using System;
using System.Diagnostics.CodeAnalysis;
using CesiumLanguageWriter;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    [TestFixture]
    public class TestYearMonthDay
    {
        /// <summary>
        /// Leap years are divisible by 4, except for years which are divisible by 100, unless also divisible by 400.
        /// </summary>
        private static bool IsLeapYear(int year)
        {
            if (year % 400 == 0)
                return true;
            if (year % 100 == 0)
                return false;
            return year % 4 == 0;
        }

        public enum Month
        {
            January = 1,
            February = 2,
            March = 3,
            April = 4,
            May = 5,
            June = 6,
            July = 7,
            August = 8,
            September = 9,
            October = 10,
            November = 11,
            December = 12,
        }

        private static int DaysInMonth(int year, Month month)
        {
            switch (month)
            {
                case Month.February:
                    return IsLeapYear(year) ? 29 : 28;
                case Month.April:
                case Month.June:
                case Month.September:
                case Month.November:
                    return 30;
                default:
                    return 31;
            }
        }

        /// <summary>
        /// Tests that an appropriate exception is thrown when constructing a
        /// YearMonthDay with an invalid date.
        /// </summary>
        [Test]
        public void TestConstructWithInvalidDate()
        {
            Assert.Throws<ArgumentException>(() =>
            {
                var unused = new YearMonthDay(2006, 2, 29);
            });
        }

        /// <summary>
        /// Tests that leap years are detected correctly.
        /// </summary>
        [Test]
        public void TestIsLeapYear()
        {
            for (int year = 1; year <= 9999; ++year)
            {
                bool expected = IsLeapYear(year);
                Assert.AreEqual(expected, YearMonthDay.IsLeapYear(year));
            }
        }

        /// <summary>
        /// Tests that leap years have 366 days instead of 365.
        /// </summary>
        [Test]
        public void TestDaysInYear()
        {
            for (int year = 1; year <= 9999; ++year)
            {
                int expected = IsLeapYear(year) ? 366 : 365;
                Assert.AreEqual(expected, YearMonthDay.DaysInYear(year));
            }
        }

        /// <summary>
        /// Tests that the length of the month is reported correctly for common years and leap years.
        /// </summary>
        [Test]
        public void TestDaysInMonth([Values] Month m)
        {
            for (int year = 1; year <= 9999; ++year)
            {
                int month = (int)m;
                int expected = DaysInMonth(year, m);
                Assert.AreEqual(expected, YearMonthDay.DaysInMonth(year, month));
            }
        }

        /// <summary>
        /// Tests that month of year and day of month ranges are validated correctly.
        /// </summary>
        [Test]
        public void TestIsValidDate([Values] Month m)
        {
            const int year = 2000;
            int month = (int)m;
            int daysInMonth = YearMonthDay.DaysInMonth(year, month);

            for (int day = 1; day <= daysInMonth; ++day)
            {
                Assert.IsTrue(YearMonthDay.IsValidDate(year, month, day));
            }

            Assert.IsFalse(YearMonthDay.IsValidDate(year, month, 0));
            Assert.IsFalse(YearMonthDay.IsValidDate(year, month, daysInMonth + 1));
        }

        [Test]
        public void TestIsValidDateWithInvalidMonth()
        {
            Assert.IsFalse(YearMonthDay.IsValidDate(2000, 0, 1));
            Assert.IsFalse(YearMonthDay.IsValidDate(2000, 13, 1));
        }

        /// <summary>
        /// Tests that initialization of and access to the structure elements is performed correctly.
        /// </summary>
        [Test]
        public void TestRetainValue()
        {
            YearMonthDay date = new YearMonthDay(2000, 1, 1);
            Assert.AreEqual(2000, date.Year);
            Assert.AreEqual(1, date.Month);
            Assert.AreEqual(1, date.Day);
        }

        /// <summary>
        /// Tests the check for EXACT equality.
        /// </summary>
        [Test]
        [SuppressMessage("Assertion", "NUnit2010", Justification = "This is specifically testing equality methods and operators")]
        public void TestEquality()
        {
            var first = new YearMonthDay(2000, 1, 1);
            var second = new YearMonthDay(2000, 1, 1);
            Assert.AreEqual(first, second);
            Assert.IsTrue(first.Equals(second));
            Assert.IsTrue(second.Equals(first));
            Assert.AreEqual(0, first.CompareTo(second));
            Assert.AreEqual(0, second.CompareTo(first));

            Assert.IsTrue(first == second);
            Assert.IsTrue(second == first);

            second = new YearMonthDay(2001, 1, 1);
            Assert.AreNotEqual(first, second);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));
            Assert.AreNotEqual(0, first.CompareTo(second));
            Assert.AreNotEqual(0, second.CompareTo(first));

            second = new YearMonthDay(2000, 2, 1);
            Assert.AreNotEqual(first, second);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));
            Assert.AreNotEqual(0, first.CompareTo(second));
            Assert.AreNotEqual(0, second.CompareTo(first));

            second = new YearMonthDay(2000, 1, 2);
            Assert.AreNotEqual(first, second);
            Assert.IsFalse(first.Equals(second));
            Assert.IsFalse(second.Equals(first));
            Assert.AreNotEqual(0, first.CompareTo(second));
            Assert.AreNotEqual(0, second.CompareTo(first));

            Assert.IsTrue(first != second);
            Assert.IsTrue(second != first);

            object obj = new YearMonthDay(2004, 2, 21);
            Assert.AreNotEqual(first, obj);

            object differentType = 5;
            Assert.AreNotEqual(first, differentType);
        }

        [Test]
        public void TestDayOfYear()
        {
            YearMonthDay nonLeapBeforeEndOfFeb = new YearMonthDay(2006, 2, 15);
            Assert.AreEqual(46, nonLeapBeforeEndOfFeb.DayOfYear);

            YearMonthDay nonLeapAfterEndOfFeb = new YearMonthDay(2006, 3, 14);
            Assert.AreEqual(73, nonLeapAfterEndOfFeb.DayOfYear);

            YearMonthDay leapBeforeEndOfFeb = new YearMonthDay(2008, 2, 15);
            Assert.AreEqual(46, leapBeforeEndOfFeb.DayOfYear);

            YearMonthDay leapAfterEndOfFeb = new YearMonthDay(2008, 3, 14);
            Assert.AreEqual(74, leapAfterEndOfFeb.DayOfYear);
        }

        /// <summary>
        /// Tests the CompareTo methods and the comparison operators.
        /// </summary>
        [Test]
        [SuppressMessage("Assertion", "NUnit2043", Justification = "This is specifically testing operator overloads")]
        public void TestComparisons()
        {
            YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
            YearMonthDay ymd2 = new YearMonthDay(2006, 3, 14);
            YearMonthDay ymd3 = new YearMonthDay(2006, 5, 26);

            Assert.IsTrue(ymd1 >= ymd2);
            Assert.IsTrue(ymd1 <= ymd2);
            Assert.AreEqual(0, ymd1.CompareTo(ymd2));
            Assert.IsTrue(ymd2 < ymd3);
            Assert.IsTrue(ymd2 <= ymd3);
            Assert.IsTrue(ymd3 > ymd2);
            Assert.IsTrue(ymd3 >= ymd2);
        }

        [Test]
        [CSToJavaExclude]
        public void TestCompareToObject()
        {
            var ymd = new YearMonthDay(2006, 3, 14);
            object ymdAsObject = new YearMonthDay(2004, 2, 21);

            Assert.Greater(ymd.CompareTo(null), 0);
            Assert.Greater(ymd.CompareTo(ymdAsObject), 0);
        }

        /// <summary>
        /// Tests that an appropriate exception is thrown when CompareTo is
        /// invoked on a type that is not a YearMonthDay.
        /// </summary>
        [Test]
        [CSToJavaExclude]
        public void TestCompareToWrongType()
        {
            YearMonthDay ymd = new YearMonthDay(2006, 3, 14);
            Assert.Throws<ArgumentException>(() =>
            {
                int unused = ymd.CompareTo(5);
            });
        }

        /// <summary>
        /// Tests that YearMonthDay.GetHashCode returns something at least reasonably random.
        /// </summary>
        [Test]
        public void TestGetHashCode()
        {
            YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
            YearMonthDay ymd2 = new YearMonthDay(2006, 3, 14);
            YearMonthDay ymd3 = new YearMonthDay(2006, 5, 26);
            Assert.AreEqual(ymd1.GetHashCode(), ymd2.GetHashCode());
            Assert.AreNotEqual(ymd1.GetHashCode(), ymd3.GetHashCode());
        }

        /// <summary>
        /// Tests the <see cref="YearMonthDay.ToString"/> method.
        /// </summary>
        [Test]
        public void TestToString()
        {
            YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
            Assert.AreEqual("2006:3:14", ymd1.ToString());
        }

        /// <summary>
        /// Tests the constructor overload that takes a <see cref="JulianDate"/>.
        /// </summary>
        [Test]
        public void TestConstructFromJulianDate()
        {
            DateTime dt = new DateTime(2008, 10, 23, 12, 0, 0);
            JulianDate jd = new JulianDate(dt);

            YearMonthDay ymd = new YearMonthDay(jd);
            Assert.AreEqual(2008, ymd.Year);
            Assert.AreEqual(10, ymd.Month);
            Assert.AreEqual(23, ymd.Day);

            dt = new DateTime(2008, 10, 23, 0, 0, 0);
            jd = new JulianDate(dt);

            ymd = new YearMonthDay(jd);
            Assert.AreEqual(2008, ymd.Year);
            Assert.AreEqual(10, ymd.Month);
            Assert.AreEqual(23, ymd.Day);

            dt = new DateTime(2008, 10, 23, 23, 59, 59);
            jd = new JulianDate(dt);

            ymd = new YearMonthDay(jd);
            Assert.AreEqual(2008, ymd.Year);
            Assert.AreEqual(10, ymd.Month);
            Assert.AreEqual(23, ymd.Day);
        }

        /// <summary>
        /// Tests the constructor overload that takes a year and the day of the year as parameters.
        /// </summary>
        [Test]
        [Combinatorial]
        public void TestConstructFromDayOfYear([Values(2000, 2001)] int year, [Values] Month m)
        {
            int month = (int)m;
            // Test each day of the month.
            for (int day = 1; day <= YearMonthDay.DaysInMonth(year, month); ++day)
            {
                int dayOfYear = new DateTime(year, month, day).DayOfYear;

                var ymd = new YearMonthDay(year, dayOfYear);
                Assert.AreEqual(year, ymd.Year);
                Assert.AreEqual(month, ymd.Month);
                Assert.AreEqual(day, ymd.Day);
            }
        }

        [Test]
        public void ConstructorThrowsWithInvalidDayOfYear()
        {
            var exception = Assert.Throws<ArgumentException>(() =>
            {
                var unused = new YearMonthDay(2000, 0);
            });
            Assert.AreEqual("dayOfYear", exception.ParamName);

            exception = Assert.Throws<ArgumentException>(() =>
            {
                var unused = new YearMonthDay(2000, 367);
            });
            Assert.AreEqual("dayOfYear", exception.ParamName);
        }

        [Test]
        public void TestJulianDayNumber()
        {
            const int julianDayNumber = 2454959;
            YearMonthDay ymd = new YearMonthDay(julianDayNumber);
            Assert.AreEqual(julianDayNumber, ymd.JulianDayNumber);
        }

        [Test]
        public void TestDayOfWeek()
        {
            YearMonthDay ymd = new YearMonthDay(2009, 06, 10);
            Assert.AreEqual(DayOfWeek.Wednesday, ymd.DayOfWeek);

            ymd = new YearMonthDay(2009, 06, 11);
            Assert.AreEqual(DayOfWeek.Thursday, ymd.DayOfWeek);
        }

        [Test]
        public void TestRoundTripDefaultConstructed()
        {
            YearMonthDay ymd = new YearMonthDay();
            YearMonthDay ymd2 = new YearMonthDay(ymd.JulianDayNumber);

            Assert.AreEqual(ymd, ymd2);
        }

        [Test]
        public void TestDefaultConstructedIsValid()
        {
            YearMonthDay ymd = new YearMonthDay();
            YearMonthDay ymd2 = new YearMonthDay(ymd.Year, ymd.Month, ymd.Day);

            Assert.AreEqual(ymd, ymd2);
            Assert.AreEqual(ymd.DayOfWeek, ymd2.DayOfWeek);
            Assert.AreEqual(ymd.DayOfYear, ymd2.DayOfYear);
        }
    }
}

File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestYearMonthDay.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.Action;
import agi.foundation.compatibility.annotations.CS2JWarning;
import agi.foundation.compatibility.ArgumentException;
import agi.foundation.compatibility.AssertHelper;
import agi.foundation.compatibility.DateTimeHelper;
import agi.foundation.compatibility.Enumeration;
import agi.foundation.compatibility.EnumHelper;
import agi.foundation.compatibility.IEquatable;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.TypeLiteral;
import cesiumlanguagewriter.*;
import java.time.DayOfWeek;
import java.time.ZonedDateTime;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.runners.MethodSorters;
import org.junit.Test;

@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestYearMonthDay {
    /**
    * Leap years are divisible by 4, except for years which are divisible by 100, unless also divisible by 400.
    */
    private static boolean isLeapYear(int year) {
        if (year % 400 == 0) {
            return true;
        }
        if (year % 100 == 0) {
            return false;
        }
        return year % 4 == 0;
    }

    public static enum Month implements Enumeration {
        JANUARY(1),
        FEBRUARY(2),
        MARCH(3),
        APRIL(4),
        MAY(5),
        JUNE(6),
        JULY(7),
        AUGUST(8),
        SEPTEMBER(9),
        OCTOBER(10),
        NOVEMBER(11),
        DECEMBER(12);

        private final int value;

        Month(int value) {
            this.value = value;
        }

        /**
        * Get the numeric value associated with this enum constant.
        * @return A numeric value.
        */
        @Override
        public int getValue() {
            return value;
        }

        /**
        * Get the enum constant that is associated with the given numeric value.
        * @return The enum constant associated with value.
        * @param value a numeric value.
        */
        @Nonnull
        public static Month getFromValue(int value) {
            switch (value) {
            case 1:
                return JANUARY;
            case 2:
                return FEBRUARY;
            case 3:
                return MARCH;
            case 4:
                return APRIL;
            case 5:
                return MAY;
            case 6:
                return JUNE;
            case 7:
                return JULY;
            case 8:
                return AUGUST;
            case 9:
                return SEPTEMBER;
            case 10:
                return OCTOBER;
            case 11:
                return NOVEMBER;
            case 12:
                return DECEMBER;
            default:
                throw new IllegalArgumentException("Undefined enum value.");
            }
        }

        /**
        * Get the enum constant that is considered to be the default.
        * @return The default enum constant.
        */
        @Nullable
        public static Month getDefault() {
            return null;
        }
    }

    private static int daysInMonth(int year, @Nonnull Month month) {
        switch (month) {
        case FEBRUARY: {
            return isLeapYear(year) ? 29 : 28;
        }
        case APRIL:
        case JUNE:
        case SEPTEMBER:
        case NOVEMBER: {
            return 30;
        }
        default: {
            return 31;
        }
        }
    }

    /**
    * Tests that an appropriate exception is thrown when constructing a
    YearMonthDay with an invalid date.
    */
    @Test
    public final void testConstructWithInvalidDate() {
        AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            YearMonthDay unused = new YearMonthDay(2006, 2, 29);
        }));
    }

    /**
    * Tests that leap years are detected correctly.
    */
    @Test
    public final void testIsLeapYear() {
        for (int year = 1; year <= 9999; ++year) {
            boolean expected = isLeapYear(year);
            AssertHelper.assertEquals(expected, YearMonthDay.isLeapYear(year));
        }
    }

    /**
    * Tests that leap years have 366 days instead of 365.
    */
    @Test
    public final void testDaysInYear() {
        for (int year = 1; year <= 9999; ++year) {
            int expected = isLeapYear(year) ? 366 : 365;
            AssertHelper.assertEquals(expected, YearMonthDay.daysInYear(year));
        }
    }

    /**
    * Tests that the length of the month is reported correctly for common years and leap years.
    */
    public final void testDaysInMonth(@Nonnull Month m) {
        for (int year = 1; year <= 9999; ++year) {
            int month = m.getValue();
            int expected = daysInMonth(year, m);
            AssertHelper.assertEquals(expected, YearMonthDay.daysInMonth(year, month));
        }
    }

    @Test
    public final void testDaysInMonth$Test() {
        for (final Month m : EnumHelper.getValues(Month.class)) {
            testDaysInMonth(m);
        }
    }

    /**
    * Tests that month of year and day of month ranges are validated correctly.
    */
    public final void testIsValidDate(@Nonnull Month m) {
        final int year = 2000;
        int month = m.getValue();
        int daysInMonth = YearMonthDay.daysInMonth(year, month);
        for (int day = 1; day <= daysInMonth; ++day) {
            Assert.assertTrue(YearMonthDay.isValidDate(year, month, day));
        }
        Assert.assertFalse(YearMonthDay.isValidDate(year, month, 0));
        Assert.assertFalse(YearMonthDay.isValidDate(year, month, daysInMonth + 1));
    }

    @Test
    public final void testIsValidDate$Test() {
        for (final Month m : EnumHelper.getValues(Month.class)) {
            testIsValidDate(m);
        }
    }

    @Test
    public final void testIsValidDateWithInvalidMonth() {
        Assert.assertFalse(YearMonthDay.isValidDate(2000, 0, 1));
        Assert.assertFalse(YearMonthDay.isValidDate(2000, 13, 1));
    }

    /**
    * Tests that initialization of and access to the structure elements is performed correctly.
    */
    @Test
    public final void testRetainValue() {
        YearMonthDay date = new YearMonthDay(2000, 1, 1);
        AssertHelper.assertEquals(2000, date.getYear());
        AssertHelper.assertEquals(1, date.getMonth());
        AssertHelper.assertEquals(1, date.getDay());
    }

    /**
    * Tests the check for EXACT equality.
    */
    @CS2JWarning("Unhandled attribute removed: SuppressMessage")
    @Test
    public final void testEquality() {
        YearMonthDay first = new YearMonthDay(2000, 1, 1);
        YearMonthDay second = new YearMonthDay(2000, 1, 1);
        AssertHelper.assertEquals(first, second);
        Assert.assertTrue(first.equalsType(second));
        Assert.assertTrue(second.equalsType(first));
        AssertHelper.assertEquals(0, first.compareTo(second));
        AssertHelper.assertEquals(0, second.compareTo(first));
        Assert.assertTrue(YearMonthDay.equals(first, second));
        Assert.assertTrue(YearMonthDay.equals(second, first));
        second = new YearMonthDay(2001, 1, 1);
        AssertHelper.assertNotEqual(first, second);
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        AssertHelper.assertNotEqual(0, first.compareTo(second));
        AssertHelper.assertNotEqual(0, second.compareTo(first));
        second = new YearMonthDay(2000, 2, 1);
        AssertHelper.assertNotEqual(first, second);
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        AssertHelper.assertNotEqual(0, first.compareTo(second));
        AssertHelper.assertNotEqual(0, second.compareTo(first));
        second = new YearMonthDay(2000, 1, 2);
        AssertHelper.assertNotEqual(first, second);
        Assert.assertFalse(first.equalsType(second));
        Assert.assertFalse(second.equalsType(first));
        AssertHelper.assertNotEqual(0, first.compareTo(second));
        AssertHelper.assertNotEqual(0, second.compareTo(first));
        Assert.assertTrue(YearMonthDay.notEquals(first, second));
        Assert.assertTrue(YearMonthDay.notEquals(second, first));
        Object obj = new YearMonthDay(2004, 2, 21);
        AssertHelper.assertNotEqual(first, obj);
        Object differentType = 5;
        AssertHelper.assertNotEqual(first, differentType);
    }

    @Test
    public final void testDayOfYear() {
        YearMonthDay nonLeapBeforeEndOfFeb = new YearMonthDay(2006, 2, 15);
        AssertHelper.assertEquals(46, nonLeapBeforeEndOfFeb.getDayOfYear());
        YearMonthDay nonLeapAfterEndOfFeb = new YearMonthDay(2006, 3, 14);
        AssertHelper.assertEquals(73, nonLeapAfterEndOfFeb.getDayOfYear());
        YearMonthDay leapBeforeEndOfFeb = new YearMonthDay(2008, 2, 15);
        AssertHelper.assertEquals(46, leapBeforeEndOfFeb.getDayOfYear());
        YearMonthDay leapAfterEndOfFeb = new YearMonthDay(2008, 3, 14);
        AssertHelper.assertEquals(74, leapAfterEndOfFeb.getDayOfYear());
    }

    /**
    * Tests the CompareTo methods and the comparison operators.
    */
    @CS2JWarning("Unhandled attribute removed: SuppressMessage")
    @Test
    public final void testComparisons() {
        YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
        YearMonthDay ymd2 = new YearMonthDay(2006, 3, 14);
        YearMonthDay ymd3 = new YearMonthDay(2006, 5, 26);
        Assert.assertTrue(YearMonthDay.greaterThanOrEqual(ymd1, ymd2));
        Assert.assertTrue(YearMonthDay.lessThanOrEqual(ymd1, ymd2));
        AssertHelper.assertEquals(0, ymd1.compareTo(ymd2));
        Assert.assertTrue(YearMonthDay.lessThan(ymd2, ymd3));
        Assert.assertTrue(YearMonthDay.lessThanOrEqual(ymd2, ymd3));
        Assert.assertTrue(YearMonthDay.greaterThan(ymd3, ymd2));
        Assert.assertTrue(YearMonthDay.greaterThanOrEqual(ymd3, ymd2));
    }

    /**
    * Tests that YearMonthDay.GetHashCode returns something at least reasonably random.
    */
    @Test
    public final void testGetHashCode() {
        YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
        YearMonthDay ymd2 = new YearMonthDay(2006, 3, 14);
        YearMonthDay ymd3 = new YearMonthDay(2006, 5, 26);
        AssertHelper.assertEquals(ymd1.hashCode(), ymd2.hashCode());
        AssertHelper.assertNotEqual(ymd1.hashCode(), ymd3.hashCode());
    }

    /**
    * Tests the {@link YearMonthDay#toString} method.
    */
    @Test
    public final void testToString() {
        YearMonthDay ymd1 = new YearMonthDay(2006, 3, 14);
        AssertHelper.assertEquals("2006:3:14", ymd1.toString());
    }

    /**
    * Tests the constructor overload that takes a {@link JulianDate}.
    */
    @Test
    public final void testConstructFromJulianDate() {
        ZonedDateTime dt = DateTimeHelper.create(2008, 10, 23, 12, 0, 0);
        JulianDate jd = new JulianDate(dt);
        YearMonthDay ymd = new YearMonthDay(jd);
        AssertHelper.assertEquals(2008, ymd.getYear());
        AssertHelper.assertEquals(10, ymd.getMonth());
        AssertHelper.assertEquals(23, ymd.getDay());
        dt = DateTimeHelper.create(2008, 10, 23, 0, 0, 0);
        jd = new JulianDate(dt);
        ymd = new YearMonthDay(jd);
        AssertHelper.assertEquals(2008, ymd.getYear());
        AssertHelper.assertEquals(10, ymd.getMonth());
        AssertHelper.assertEquals(23, ymd.getDay());
        dt = DateTimeHelper.create(2008, 10, 23, 23, 59, 59);
        jd = new JulianDate(dt);
        ymd = new YearMonthDay(jd);
        AssertHelper.assertEquals(2008, ymd.getYear());
        AssertHelper.assertEquals(10, ymd.getMonth());
        AssertHelper.assertEquals(23, ymd.getDay());
    }

    /**
    * Tests the constructor overload that takes a year and the day of the year as parameters.
    */
    public final void testConstructFromDayOfYear(int year, @Nonnull Month m) {
        int month = m.getValue();
        // Test each day of the month.
        for (int day = 1; day <= YearMonthDay.daysInMonth(year, month); ++day) {
            int dayOfYear = DateTimeHelper.create(year, month, day).getDayOfYear();
            YearMonthDay ymd = new YearMonthDay(year, dayOfYear);
            AssertHelper.assertEquals(year, ymd.getYear());
            AssertHelper.assertEquals(month, ymd.getMonth());
            AssertHelper.assertEquals(day, ymd.getDay());
        }
    }

    @Test
    public final void testConstructFromDayOfYear$Test() {
        for (final int year : new int[] {
            2000,
            2001
        }) {
            for (final Month m : EnumHelper.getValues(Month.class)) {
                testConstructFromDayOfYear(year, m);
            }
        }
    }

    @Test
    public final void constructorThrowsWithInvalidDayOfYear() {
        ArgumentException exception = AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            YearMonthDay unused = new YearMonthDay(2000, 0);
        }));
        AssertHelper.assertEquals("dayOfYear", exception.getParamName());
        exception = AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            YearMonthDay unused = new YearMonthDay(2000, 367);
        }));
        AssertHelper.assertEquals("dayOfYear", exception.getParamName());
    }

    @Test
    public final void testJulianDayNumber() {
        final int julianDayNumber = 2454959;
        YearMonthDay ymd = new YearMonthDay(julianDayNumber);
        AssertHelper.assertEquals(julianDayNumber, ymd.getJulianDayNumber());
    }

    @Test
    public final void testDayOfWeek() {
        YearMonthDay ymd = new YearMonthDay(2009, 6, 10);
        AssertHelper.assertEquals(DayOfWeek.WEDNESDAY, ymd.getDayOfWeek());
        ymd = new YearMonthDay(2009, 6, 11);
        AssertHelper.assertEquals(DayOfWeek.THURSDAY, ymd.getDayOfWeek());
    }

    @Test
    public final void testRoundTripDefaultConstructed() {
        YearMonthDay ymd = new YearMonthDay();
        YearMonthDay ymd2 = new YearMonthDay(ymd.getJulianDayNumber());
        AssertHelper.assertEquals(ymd, ymd2);
    }

    @Test
    public final void testDefaultConstructedIsValid() {
        YearMonthDay ymd = new YearMonthDay();
        YearMonthDay ymd2 = new YearMonthDay(ymd.getYear(), ymd.getMonth(), ymd.getDay());
        AssertHelper.assertEquals(ymd, ymd2);
        AssertHelper.assertEquals(ymd.getDayOfWeek(), ymd2.getDayOfWeek());
        AssertHelper.assertEquals(ymd.getDayOfYear(), ymd2.getDayOfYear());
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }
}